src/python/pants/backend/go/util_rules/compile.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+from pants.backend.go.util_rules.sdk import GoSdkProcess
+from pants.engine.fs import Digest
+from pants.engine.process import ProcessResult
+from pants.engine.rules import Get, collect_rules, rule
+
+
+@dataclass(frozen=True)
+class CompileGoSourcesRequest:
+    """Compile Go sources into the package archive __pkg__.a."""
+
+    # The `Digest` containing the input Go source files and input packages.
+    digest: Digest
+
+    # Paths to each source file to compile.
+    sources: tuple[str, ...]
+
+    # The import path for the package being compiled.
+    import_path: str
+
+    # Description to use for the compilation.
+    description: str
+
+    # Import configuration
+    import_config_path: str | None = None
+
+    # Optional symabis file to use.
+    symabis_path: str | None = None
+
+
+@dataclass(frozen=True)
+class CompiledGoSources:
```

Comment:
Not blocking, but consider making this optionally fallible (á la #12859) to make it easier to add `check` support.

Reply:
That support should go on the rules for `BuildGoPackageRequest` and `BuiltGoPackage` which is what `check` would use. Some Go packages have assembly so merely calling `CompileGoSourcesRequest` would be insufficient to build those packages.